### ISSUE TITLE
 Failed to find implementation for task with name,  Execution will be excluded from due. Either delete the execution from the database, or add an implementation for it

### ISSUE THREAD

**Issue by ashishkattamuri** [kagkarlsson/db-scheduler#233]
I am seeing below error when I schedule a list of recurring tasks ( around 20 recurring tasks scheduled). The scheduler randomly throws below warning and exclude one or two of the tasks from execution and delete after the default period. I am using the below code to get recurring task and pass it to the scheduler. From the code, I see the startTasks method is adding all the tasks to knownTasks, but I am not sure why its complaining the tasks as unknown.

"Failed to find implementation for task with name 'testTask'. Execution will be excluded from due. Either delete the execution from the database, or add an implementation for it. The scheduler may be configured to automatically delete unresolved tasks after a certain period of time."

```
        final Scheduler scheduler = Scheduler.create(dataSource).startTasks(recurringTasks).threads(recurringTasks.size())
                .pollingInterval(Duration.ofSeconds(10))
                .heartbeatInterval(Duration.ofSeconds(10))
                .registerShutdownHook()
                .build();
        taskScheduler = scheduler;
        scheduler.start();


private RecurringTask<Void> getRecurringTask(Task task, TaskConfig config)
    {
        return Tasks.recurring(task.getType().name(), toSchedule(config))
                .onFailure(new OnFailureReschedule(FixedDelay.ofSeconds(60)))
                .execute((VoidExecutionHandler<Void>) task);
    }
```

**Comment by ashishkattamuri** [comment]
I have tested that scenario of old executions in the database as well. It make sense to see this error on old executions, but I am seeing this errors on recurrent and active tasks as well and all the tasks implement VoidExecutionHandler<Void>.

We had to frequently restart our nodes in case any of the active and implemented tasks are getting deregistered. 
At the restart of the nodes, I see all the tasks ( total of 35 tasks) in the scheduled_tasks and I see the random tasks getting deleted from scheduled_tasks. 

I see that we are passing all tasks to the startTasks method, but still seeing some tasks not seen as knownTasks.

**Comment by kagkarlsson** [comment]
Closing this question, I hope you found the cause of the problem

**Comment by hana-krizova** [comment]
Good morning,

thank you very much for the db-scheduler, I like it very much.

I get the same error when I use multiple instances of Scheduler in one application. Then the wrong TaskResolver (and probably JdbcTaskRepository) is sometimes selected, which incorrectly marks the task unresolved at line 61 in the TaskResolver.java class

Is it allowed to use multiple instances of Scheduler in one application at all?

**Comment by ashishkattamuri** [comment]
> Closing this question, I hope you found the cause of the problem

Was able to resolve by backing it by different tables.

**Comment by hana-krizova** [comment]
I have a use case, when users can add schedulers manually.

But thanks, I can live with one Scheduler instance or with more tables.

**Comment by kagkarlsson** [comment]
I think you are doing it wrong. Users should not trigger creation of new Scheduler-instances, they should trigger the scheduling of new instances of a specific task. For that case, you only need one Scheduler-instance..

**Comment by hana-krizova** [comment]
Thank you, Mr. Karlsson,

my use case is a bit special: Users can any time write their own scripts and dynamically ask to execute them periodically. In Scheduler terminilogy - I need to add new recurring Task dynamically?

But this situation occurs so rarely that I can call Scheduler#stop() and then create a new scheduler by Scheduler#build() with new list of recurring tasks. Is it OK?

**Comment by kagkarlsson** [comment]
I think you need to do something like this: 
https://github.com/kagkarlsson/db-scheduler/blob/master/examples/features/src/main/java/com/github/kagkarlsson/examples/RecurringTaskWithPersistentScheduleMain.java#L57

1. Create a generic task that executes the scripts, start the scheduler with that
2. When users schedules a script, create a new instance of the task: 
```
scheduler.schedule(task.schedulableInstance("user-task-1", new PlainScheduleAndData(Schedules.fixedDelay(Duration.ofSeconds(1)))));
```

(though using a fitting schedule)

**Comment by hana-krizova** [comment]
Yes, I tried it, it works! Thank you :-)

But my solution - to stop Scheduler and build another one - is OK too?

I use Kotlin, I don't have Java serialization to serialize custom implementations of com.github.kagkarlsson.scheduler.task.schedule.Schedule interfaces.

Yes, I can have own serializer, but It seems to me more straightforward.

**Comment by kagkarlsson** [comment]
> But my solution - to stop Scheduler and build another one - is OK too?

Not really. I think you can run into problems doing that

I would like to provide a Kotlin serializer out of the box. I have not looked at it yet, but serialization is the feature I am currently working on

**Comment by kagkarlsson** [comment]
🎉 This issue has been resolved in `v16.9.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v16.9.0))
